subsys/net/lib/dns/dns_sd.c

```diff
@@ -0,0 +1,786 @@
+
+	rdata = (struct dns_srv_rdata *)&buf[offset];
+	rdata->priority = 0;
+	rdata->weight = 0;
+	rdata->port = htons(inst->port);
+	offset += sizeof(*rdata);
+
+	*host_offset = offset;
+
+	label_size = strlen(inst->instance);
+	buf[offset++] = label_size;
+	memcpy(&buf[offset], inst->instance, label_size);
+	offset += label_size;
+
+	domain_offset |= DNS_SD_PTR_MASK;
+	domain_offset = htons(domain_offset);
+	memcpy(&buf[offset], &domain_offset, sizeof(domain_offset));
+	offset += sizeof(domain_offset);
+
+	__ASSERT_NO_MSG(total_size == offset - buf_offset);
+
+	return offset - buf_offset;
+}
+
+int dns_sd_handle_ptr_query(const struct dns_sd_rec *inst, bool ipv6,
+			    const uint8_t *addr,
+			    uint8_t *buf, uint16_t buf_size)
+{
+	/*
+	 * RFC 6763 Section 12.1
+	 *
+	 * When including a DNS-SD Service Instance Enumeration or Selective
+	 * Instance Enumeration (subtype) PTR record in a response packet, the
+	 * server/responder SHOULD include the following additional records:
+	 *
+	 * o  The SRV record(s) named in the PTR rdata.
+	 * o  The TXT record(s) named in the PTR rdata.
+	 * o  All address records (type "A" and "AAAA") named in the SRV rdata.
+	 *	contain the SRV record(s), the TXT record(s), and the address
+	 *      records (A or AAAA)
+	 */
+
+	uint16_t instance_offset = -1;
+	uint16_t service_offset = -1;
+	uint16_t domain_offset = -1;
+	uint16_t host_offset = -1;
+	uint16_t offset = sizeof(struct dns_header);
+	struct dns_header *rsp = (struct dns_header *)buf;
+	int r;
+
+	memset(rsp, 0, sizeof(*rsp));
+
+	/* first add the answer record */
+	r = add_ptr_record(inst, DNS_SD_PTR_TTL, buf, offset,
+			   buf_size - offset,
+			   &service_offset, &instance_offset,
+			   &domain_offset);
+	if (r < 0) {
+		return r; /* LCOV_EXCL_LINE */
+	}
+
+	rsp->ancount++;
+	offset += r;
+
+	/* then add the additional records */
+	r = add_txt_record(inst, DNS_SD_TXT_TTL, instance_offset, buf,
+			   offset,
+			   buf_size - offset);
+	if (r < 0) {
+		return r; /* LCOV_EXCL_LINE */
+	}
+
+	rsp->arcount++;
+	offset += r;
+
+	r = add_srv_record(inst, DNS_SD_SRV_TTL, instance_offset,
+			   domain_offset,
+			   buf, offset, buf_size - offset, &host_offset);
+	if (r < 0) {
+		return r; /* LCOV_EXCL_LINE */
+	}
+
+	rsp->arcount++;
+	offset += r;
+
+	if (ipv6) {
+		r = add_aaaa_record(inst, DNS_SD_AAAA_TTL, host_offset,
+				    addr,
+				    buf, offset,
+				    buf_size - offset);     /* LCOV_EXCL_LINE */
+	} else {
+		r = add_a_record(inst, DNS_SD_A_TTL, host_offset,
+				 *((uint32_t *)addr), buf, offset,
+				 buf_size - offset);
+	}
+
+	if (r < 0) {
+		return r; /* LCOV_EXCL_LINE */
+	}
+
+	rsp->arcount++;
+	offset += r;
+
+	rsp->flags = htons(BIT(15) | BIT(10));
+	rsp->ancount = htons(rsp->ancount);
+	rsp->arcount = htons(rsp->arcount);
+
+	return offset;
+}
+
+/* TODO: dns_sd_handle_srv_query() */
+/* TODO: dns_sd_handle_txt_query() */
+
+bool dns_sd_rec_match(const struct dns_sd_rec *a, struct dns_sd_rec *b)
+{
+	#define WILDCARD NULL
```

Comment:
Please put the define either to the beginning of the file or outside of this function, and do not indent